### PR TITLE
fix: Transactions tag and data serialization

### PR DESCRIPTION
**Pull request by AlexanderDubikov**
Fix transaction serialization #1311
Also [added](https://github.com/getsentry/sentry-cocoa/commit/e8e0fefa494d23490164af2ecc9389591b53d8a0) trace context tags to transaction level during serialisation

## :scroll: Description

- SentryTransaction uses SentryTracer instead of SentrySpan to store time series data.
- SentryTracer stores tags and data itself. It was rootSpan's responsibility before.
- Added trace context tags to upper level of serialised transaction
- Proceeded using [these](https://github.com/getsentry/sentry-cocoa/pull/1703) thoughts.
- This is a [1700s](https://github.com/getsentry/sentry-cocoa/pull/1700) copy. Couldn't reopen it.

## :bulb: Motivation and Context

While writing network transactions monitoring I found out that a transaction can't add any tags. 
Comparing with Android `envelope` json I understood that this thing is an issue. This is the [link](https://github.com/getsentry/sentry-cocoa/issues/1311). 

So the main problem was that `tags` and `data` were kept in the `trace` dictionary at the resulting `envelope`. 

I also found that `trace context tags` remain in the `trace` dictionary inside the envelope. Which also doesn't work.

## :green_heart: How did you test it?

- Tags appear at the event columns at the dashboards. JSON structure for the envelope event matches Android SDK. 
- I also added some tests for SentryTracer, SentryClient.
- Removed some test checks at SentrySpanTests testSerialization() method. 
- Added 5 tests to SentryTransactionTests

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed (added some links to the issue and a small description)
- [x] Review from the native team if needed
- [x] No breaking changes

## :crystal_ball: Next steps

I can remove those `given-when-then` comments used in tests.
The  `trace context tags` fix can also be reverted.

### CODE REVIEW

File: Tests/SentryTests/Transaction/SentrySpanTests.swift [hunk 1]
```diff
@@ -140,8 +140,6 @@ class SentrySpanTests: XCTestCase {
     func testSerialization() {
         let span = fixture.getSut()
         
-        span.setExtra(value: fixture.extraValue, key: fixture.extraKey)
-        span.setTag(value: fixture.extraValue, key: fixture.extraKey)
```
Comment (philipphofmann) [inline Tests/SentryTests/Transaction/SentrySpanTests.swift:144]: `h`: I think we should not remove this here. This test should still work.

File: Tests/SentryTests/SentryClientTests.swift [hunk 2]
```diff
@@ -169,6 +169,27 @@ class SentryClientTest: XCTestCase {
         }
     }
     
+    func testCaptureEventWithScope_SerializedTagsAndExtraShouldMatch() {
```
Comment (philipphofmann) [inline Tests/SentryTests/SentryClientTests.swift:172]: I wonder why did you add this test? This should already be tested somewhere.

File: Tests/SentryTests/Transaction/SentryTransactionTests.swift [hunk 3]
```diff
@@ -3,20 +3,59 @@ import XCTest
...
+            let client = TestClient(options: Options())!
+            client.options.tracesSampleRate = 1
+            
+            let hub = TestHub(client: client, andScope: scope)
+            let trace = SentryTracer(transactionContext: self.getContext(), hub: hub)
+            let transaction = Transaction(trace: trace, children: [])
+            return transaction
+        }
+        
+        init() { }
```
Comment (philipphofmann) [inline Tests/SentryTests/Transaction/SentryTransactionTests.swift:40]:
```suggestion
```

**Review by philipphofmann — COMMENTED** [review]
First pass. Didn't fully finish yet. Thanks for doing this @AlexanderDubikov

File: Tests/SentryTests/SentryClientTests.swift (hunk 2 above)
Comment (AlexanderDubikov) [reply Tests/SentryTests/SentryClientTests.swift:172]: If this was tested before,  the [1311](https://github.com/getsentry/sentry-cocoa/issues/1311) issue won't happen :) 

This is somehow an integration test. It considers Event, Transaction, Scope. The logic for this thing is inside this method.

`- (SentryId *)sendEvent:(SentryEvent *)event
                  withScope:(SentryScope *)scope
     alwaysAttachStacktrace:(BOOL)alwaysAttachStacktrace
               isCrashEvent:(BOOL)isCrashEvent
    additionalEnvelopeItems:(NSArray<SentryEnvelopeItem *> *)additionalEnvelopeItems`
     
   `[self getTraceStateWithEvent:event withScope:scope]` - this part.
   
   Maybe I'm getting it wrong, but this is SentryClient's logic.

File: Tests/SentryTests/Transaction/SentrySpanTests.swift (hunk 1 above)
Comment (AlexanderDubikov) [reply Tests/SentryTests/Transaction/SentrySpanTests.swift]: Added some extra logic to SentryTracer class. It didn't have its own tags and data before, but they were lost during serialization method (because SentryTracer was using its root span's ones)

File: Tests/SentryTests/Transaction/SentryTransactionTests.swift [hunk 4]
```diff
@@ -36,4 +73,84 @@ class SentryTransactionTests: XCTestCase {
         let actualMeasurements = actual["measurements"] as? [String: [String: String]]
         XCTAssertEqual(["value": self.description], actualMeasurements?["app_start_cold"] )
     }
+    
+    func testSerialize_Tags() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Transaction/SentryTransactionTests.swift:77]: Nice tests, thanks

**Review by philipphofmann — APPROVED** [review]
LGTM 🚀 Thanks for your patience @AlexanderDubikov
